docs/_posts/2019-11-14-wash-gem.md

```diff
@@ -0,0 +1,160 @@
+---
+title: "Using the wash gem to write your external plugins"
+description: "Shows how the wash gem can be used to write external plugins"
+author: enisinan
+---
+
+Wash ships with plugins that let you interact with your AWS, GCP, Docker, and Kubernetes resources. You can also write your own plugins via the [external plugin interface]({{ '/docs/external-plugins' | relative_url }}). An external plugin consists of a *plugin* script. Wash shells out to this script whenever it needs to invoke an entry’s supported action (like *list*), or if it needs to query something about the entry (like its metadata). In a more general sense, Wash shells out to the plugin script whenever it needs to invoke an entry’s supported *method*. All plugin script invocations adopt the following calling convention
+
+```
+<plugin_script> <method> <path> <state> <args...>
+```
+
+where
+
+* `<method>` is the Wash method that’s being invoked. This includes Wash actions like `list` and `exec`, and also non-Wash actions like `schema` and `metadata`.
+
+* `<path>` is the entry’s filesystem path rooted at Wash’s mountpoint. For example, `/my_plugin` would be the `my_plugin`'s plugin root. `/my_plugin/foo` would be the `foo` child of the `my_plugin` entry.
+
+* `<state>` consists of the minimum amount of information required to reconstruct the entry inside the plugin. It can be _any_ string. For example, `'{"klass": "Root"}'` could be a JSON object representing the plugin root in a higher-level programming language like Python or Ruby.
+
+* `<args...>` are `<method>`’s arguments. For example if `<method>` is `exec`, then the exec’ed command would be included in `<args...>`.
+
+Now let `<entry> = <path> <state>`. Then the plugin script’s usage becomes
+
+```
+<plugin_script> <method> <entry> <args...>
+```
+
+From this usage, we see that `<path>` and `<state>` are two different representations of an entry. `<path>` is useful for simple plugins where reconstructing the entry is easy. `<state>` is useful for more complicated plugins where entries could be represented as classes.
+
+If you're a Ruby developer, then the [wash gem](https://github.com/puppetlabs/wash-ruby/tree/0.1.2) makes external plugin development easy by letting you model your entries as classes. We'll show you what this looks like by walking you through some of the [puppetwash](https://github.com/puppetlabs/puppetwash/tree/dadfbc3bb7b4f8fe379086182c3006237028ffc8) plugin's implementation, specifically the [puppetwash.rb](https://github.com/puppetlabs/puppetwash/blob/dadfbc3bb7b4f8fe379086182c3006237028ffc8/puppetwash.rb) file.
```

Comment:
This isn't a tagged version. Were you going to tag it?

Reply:
Is tagging all we need to release it?

Reply:
Yeah, I think so.